### PR TITLE
Update pattern-matching todo lists.

### LINKED ISSUE #7703: Two small changes to the pattern matching spec
Per https://github.com/dotnet/roslyn/issues/206#issuecomment-167268457

The use of a pattern in a `case` was intended to be an *extension* to the existing alternative of a constant expression. That should be clarified, though. In addition, it was intended that a *case-guard* could be used with a constant-valued `case`. I will amend the [spec](https://github.com/dotnet/roslyn/blob/future/docs/features/patterns.md) to clarify.

The `goto case` statement would be relaxed in two ways. First, the expression no longer needs to be a constant. Second, the expression no longer needs to correspond directly to a `case` label. Logically the `goto case` statement would take an expression and dispatches the whole `switch` statement with the new expression. This needs to be added to the [spec](https://github.com/dotnet/roslyn/blob/future/docs/features/patterns.md) too.

### LINKED ISSUE #9645: GenerateNullCoalescingBadBinaryOpsError crashes
**Version Used**: current future branch

**Steps to Reproduce**:

`StringBuilder sb = null ?? throw new ArgumentNullException();`

**Actual Behavior**: (BoundThrowException)rightOperand.Display == null in GenerateNullCoalescingBadBinaryOpsError raises Debug.Assert in AssertMessageSerializable

### PR DESCRIPTION
**Pull request by gafter**
Add a number of PROTOTYPE comments.
Fix formatting of throw expressions in diagnostics.
Update pattern-matching specification and work list.
Closes #7703
Fixes #9645

@dotnet/roslyn-compiler @AlekseyTs Please review.

### CODE REVIEW

**Comment by jaredpar** [comment]
:+1:

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs [hunk 1]
```diff
@@ -192,19 +192,7 @@ private TypeSymbol GetAnonymousTypeFieldType(BoundExpression expression, CSharpS
...
-                    else if (expression.Kind == BoundKind.MethodGroup)
-                    {
-                        errorArg = MessageID.IDS_MethodGroup.Localize();
-                    }
-                    else
-                    {
-                        Debug.Assert(expression.IsLiteralNull(), "How did we successfully bind an expression without a type?");
-                        errorArg = MessageID.IDS_NULL.Localize();
-                    }
+                    errorArg = expression.Display;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs]: Is this change related to the feature work? If not, we might want to put it in the 'future' on its own. Please create an issue for this.

File: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs [hunk 2]
```diff
@@ -568,21 +568,7 @@ private void ReduceLet(LetClauseSyntax let, QueryTranslationState state, Diagnos
...
-                    {
-                        id = MessageID.IDS_MethodGroup;
-                    }
-                    else
-                    {
-                        Debug.Assert(yExpression.IsLiteralNull(), "How did we successfully bind an expression without a type?");
-                    }
-
-                    Error(d, ErrorCode.ERR_QueryRangeVariableAssignedBadValue, errorLocation, id.Localize());
+                    Error(d, ErrorCode.ERR_QueryRangeVariableAssignedBadValue, errorLocation, yExpression.Display);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Query.cs]: Same question/suggestion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 3]
```diff
@@ -692,22 +692,8 @@ private TypeSymbol BindVariableType(CSharpSyntaxNode declarationNode, Diagnostic
...
-                    id = MessageID.IDS_MethodGroup;
-                }
-                else
-                {
-                    Debug.Assert(expression.IsLiteralNull(), "How did we successfully bind an expression without a type?");
-                }
-
                 // Cannot assign {0} to an implicitly-typed local variable
-                Error(diagnostics, ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, errorSyntax, id.Localize());
+                Error(diagnostics, ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, errorSyntax, expression.Display);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Same question/suggestion.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 4]
```diff
@@ -7785,5 +7785,26 @@ static bool Dummy(bool x, object y, object z)
...
+{
+    static void Main(string[] args)
+    {
+        var y = null ?? throw null;
+    }
+}
+";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: patternParseOptions);
+            compilation.VerifyDiagnostics(
+                // (6,17): error CS0019: Operator '??' cannot be applied to operands of type '<null>' and '<throw>'
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: The message sounds strange, it suggests that ```<throw>``` refers to a type.

**Comment by AlekseyTs** [comment]
There are three places where behavior around error reporting changed in a significant way, affecting scenarios unrelated to pattern matching. I would expect to see more testing, for all three places, specifically covering lambda expressions and method groups.

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs [hunk 5]
```diff
@@ -14,7 +15,11 @@ internal abstract partial class BoundExpression
         /// </summary>
         public virtual object Display
         {
-            get { return this.Type; }
+            get
+            {
+                Debug.Assert((object)this.Type != null, $"Unexpected null type in {this.GetType().Name}");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: null is never a good argument for a diagnostics. While it is good to have an assert, I think we should also add a fallback to something non-null.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: Yes, the same for <null>, method groups, lambdas, etc. It is the compiler's existing style.

File: src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_AnonymousTypes.cs]: It is related to the feature work, as I don't want to have to add a case for `BoundKind.ThrowExpression` everywhere this code appears. `expression.Display` was previously added for just this purpose, it just wasn't rolled out to all the sites where it should have been.

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: This returns the type whenever the type is non-null. The type can only be null for certain kinds of nodes, and all of those nodes have their own override of the `Display` property.

**Comment by gafter** [comment]
@AlekseyTs 

> There are three places where behavior around error reporting changed in a significant way, affecting scenarios unrelated to pattern matching.

We already have tests covering those existing scenarios. If you look at the structure of the changed code, you'll see that the behavior is actually the same, it is just calling a common virtual method that has the same behavior as the deleted if-then-else in each case.

**Pull request merged**